pkg/webhooks/fleet_webhook.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+)
+
+var _ webhook.CustomValidator = &FleetWebhook{}
+
+// Define a map to store mutex for each namespace
+var nsLocks = make(map[string]*sync.Mutex)
+
+// Valid cluster kinds
+var validClusterKinds = []string{"Cluster", "AttachedCluster", "CustomCluster"}
+
+type FleetWebhook struct {
+	Client client.Reader
+}
+
+func (wh *FleetWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&v1alpha1.Fleet{}).
+		WithValidator(wh).
+		Complete()
+}
+
+func (wh *FleetWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	in, ok := obj.(*v1alpha1.Fleet)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Fleet but got a %T", obj))
+	}
+
+	// Ensure only one Fleet instance in a namespace
+	mutex := getOrCreateMutexForNamespace(in.Namespace)
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	// Check if Fleet instance already exists in the namespace
+	existing := &v1alpha1.Fleet{}
+	if err := wh.Client.Get(ctx, client.ObjectKey{Namespace: in.Namespace, Name: in.Name}, existing); err == nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("a Fleet instance already exists in namespace %s: %s", existing.Namespace, existing.Name))
+	}
+
+	return nil
+}
+
+// Utility function to get or create a mutex for a namespace
+func getOrCreateMutexForNamespace(ns string) *sync.Mutex {
+	if _, exists := nsLocks[ns]; !exists {
+		nsLocks[ns] = &sync.Mutex{}
+	}
+	return nsLocks[ns]
+}
+
+func (wh *FleetWebhook) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) error {
```

Comment:
For update, we should prebent some fields update we do not support, this need to be reviewed one by one